### PR TITLE
Implement `Functors.functor`

### PR DESCRIPTION
**Pull request by devmotion**
This PR is an alternative to https://github.com/TuringLang/Bijectors.jl/pull/161. As mentioned there, `Flux.@treelike` does not exist anymore. Instead, Flux relies on `Functors.functor` and reexports `Flux.@functor`.

This PR implements `Functors.functor` for bijectors and normalizing flows. Thus users can just call `Flux.params` without implementing `@functor` for the different layers.

A dependency on the lightweight package Functors.jl is added and the optional dependency on Flux is dropped.

### CODE REVIEW

File: test/bijectors/utils.jl [hunk 1]
```diff
@@ -192,3 +192,10 @@ function test_logabsdetjac(b::Bijector{0}, xs::AbstractVector; tol=1e-6)
         @test mean(logabsdetjac(b, xs) - logjac_ad) ≤ tol
     end
 end
+
+# Check if `Functors.functor` works properly
+function test_functor(x, xs)
+    _xs, re = Functors.functor(x)
+    @test x == re(_xs)
+    @test _xs == xs
+end
```
Comment (torfjelde) [inline test/bijectors/utils.jl:201]: Is this being used anywhere?

**Review by torfjelde — CHANGES_REQUESTED** [review]
This is great stuff!
I have one comment about the testing; it seems like we currently only test a couple of the bijectors?

File: test/bijectors/utils.jl (hunk 1 above)
Comment (devmotion) [reply test/bijectors/utils.jl:201]: Yes, in the `norm_flows` tests.

**Comment by devmotion** [comment]
Yes, indeed, so far only `InvertibleBatchnorm`, `PlanarLayer`, and `RadialLayer` are tested in `norm_flows.jl`. I am still a bit confused about the test setup - where would be the best place to add tests for other bijectors? Initially I assume that most interface tests would be handled by `test_bijector` but it seems it is only used in one single test?

**Comment by torfjelde** [comment]
> where would be the best place to add tests for other bijectors? Initially I assume that most interface tests would be handled by test_bijector but it seems it is only used in one single test?

That's correct, but the idea is to start using that approach for all bijectors. So feel free to add it there for now :)

File: test/bijectors/utils.jl (hunk 1 above)
Comment (torfjelde) [reply test/bijectors/utils.jl:201]: Ah, missed that :+1:

**Comment by torfjelde** [comment]
Actually, I guess it's somewhat non-trivial to add this to `test_bijector`, as we don't have access to the parameters of the bijector itself?

**Comment by devmotion** [comment]
I assume one would have to specify the parameters explicitly, similar to `test_functor`. However, one could set to `nothing` by default and skip the tests in this case.

**Comment by devmotion** [comment]
This will lead to a stackoverflow error. But in principle this is the idea.

**Comment by torfjelde** [comment]
Oh yeah, rather:

``` julia
function test_bijector(
    b::Bijector,
    xs_true,
    ys_true,
    logjacs_true
    params;
    kwargs...
)
    test_functor(b, params)
    return test_bijector(b, xs_true, ys_true, logjacs_true; kwargs...)
end
```

**Review by torfjelde — APPROVED** [review]
I'm happy for now, and we can make it testing the functor functionality part of the tests once we unify testing a bit.

**Comment by devmotion** [comment]
Ah sorry, I forgot about the PR. OK, let's update the tests separately then!

**Pull request merged**